src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -31,31 +38,61 @@ object RuleUtils {
    *
    * @param indexManager indexManager
    * @param plan logical plan
+   * @param spark Spark session
    * @return indexes built for this plan
    */
-  def getCandidateIndexes(indexManager: IndexManager, plan: LogicalPlan): Seq[IndexLogEntry] = {
+  def getCandidateIndexes(
+      spark: SparkSession,
+      indexManager: IndexManager,
+      plan: LogicalPlan): Seq[IndexLogEntry] = {
     // Map of a signature provider to a signature generated for the given plan.
-    val signatureMap = mutable.Map[String, Option[String]]()
+    val signatureMap = mutable.Map[(String, Option[Set[Path]]), Option[String]]()
+    val hybridScanEnabled = spark.sessionState.conf
+      .getConfString(
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED,
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED_DEFAULT.toString)
+      .toBoolean
 
-    def signatureValid(entry: IndexLogEntry): Boolean = {
+    def signatureValid(entry: IndexLogEntry, sourceFileSet: Option[Set[Path]]): Boolean = {
       val sourcePlanSignatures = entry.source.plan.properties.fingerprint.properties.signatures
       assert(sourcePlanSignatures.length == 1)
       val sourcePlanSignature = sourcePlanSignatures.head
       signatureMap.getOrElseUpdate(
-        sourcePlanSignature.provider,
+        (sourcePlanSignature.provider, sourceFileSet),
         LogicalPlanSignatureProvider
           .create(sourcePlanSignature.provider)
-          .signature(plan)) match {
+          .signature(plan, sourceFileSet)) match {
         case Some(s) => s.equals(sourcePlanSignature.value)
         case None => false
       }
     }
 
+    def isCandidate(entry: IndexLogEntry): Boolean = {
+      if (hybridScanEnabled) {
+        assert(entry.relations.length == 1)
+        // entry.allSourceFiles
+        var files = scala.collection.mutable.Seq[FileInfo]()
+        plan.foreach {
+          case LogicalRelation(
+              HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+              _,
+              _,
+              _) =>
+            files ++= location.allFiles.map(f =>
+              FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+        }
+        val commonFiles = files.toSet.intersect(entry.allSourceFiles)
+        commonFiles.nonEmpty
```

Comment:
With hybrid scan, the condition for candidates would be just any common source files.
Maybe it's good to have some some threshold (like delta is less than 50%) here..?